### DISCUSSION TITLE
xmake怎么静态编译？

### DISCUSSION THREAD

**Discussion by zhangxichang** [xmake-io/xmake#2489] [Q&A]
同一个项目代码和库，我用cmake可以静态编译。但是用xmake静态编译虽然没有报错。但是二进制文件实际上没有静态编译。

这是我xmake的lua
{
set_project("GameEngine")
set_version("0.1.0")

--Vulkan
add_includedirs("./SDK/Vulkan/Include")
add_linkdirs("./SDK/Vulkan/Lib")
--GLFW
add_includedirs("./SDK/GLFW/include")
add_linkdirs("./SDK/GLFW/lib-mingw-w64")
--GLM
add_includedirs("./SDK/GLM/include")
--STB
add_includedirs("./SDK/STB/include")
--Assimp
add_includedirs("./SDK/Assimp/include")
add_linkdirs("./SDK/Assimp/lib")

target("Demo_WinDBG")
    --平台架构
    set_toolchains("mingw")
    set_plat("windows")
    set_arch("x64")

    --编译器配置
    set_languages("clatest","c++latest")
    set_optimize("fastest")
    add_cxflags("-static")

    --链接库
    add_links("vulkan-1","glfw3","assimp","zlibstatic")
    --头文件
    add_includedirs("./include")
    --源文件
    add_files("./source/*.cpp")

    --目标类型和主函数文件
    set_kind("binary")
    add_files("main.cpp")
target_end()
}

**Comment by waruqi** [comment]
不懂啥叫静态编译，只有静态链接，从没听说过，编译还能静态

> add_cxflags("-static")

-static 是 linker flags，设置到编译 flags 当然不会生效，用 add_ldflags

**Reply by zhangxichang** [reply]
问这个问题的时候我其实已经尝试过给链接器-static或者给编译器-static（问这个问题的时候是给编译器-static的状态），但是都不行，编译通过，但是文件很明显不是静态编译的。然后第二天问题就没了。这是xmake的玄学吗？
我就是因为给链接器或者给编译器甚至都给“-static”都没办法静态编译，所以我才来问的。我在想，xmake莫非是有什么特殊方便的编译机巧吗？再加上就是网上xmake的资料不能说非常丰富，只能说是根本没有。

**Reply by waruqi** [reply]
没有，只需要 `add_ldflags("-static", {force = true})` 。。你可以自己 `xmake -v` 查看完整 link flags，确认是否传入 linker 。。如果已经传入 linker ，还是没有生效，那就不是 xmake 的问题，即使是玄学，也是 gcc/clang 的玄学，你得找他们去。

**Reply by zhangxichang** [reply]
没事，反正问题莫名其妙的没了。也没有再出现了。可能是我之前的环境配置问题导致的，都有可能。环境变量也是玄学的一种。
